src/coffea/dataset_tools/filespec.py

```diff
@@ -0,0 +1,385 @@
+            except Exception:
+                pass
+        return self
+
+
+class DatasetSpec(BaseModel):
+    files: CoffeaFileDict
+    metadata: dict[Hashable, Any] | None = None
+    format: str | None = None
+    form: str | None = None
+
+    @model_validator(mode="before")
+    def preprocess_data(cls, data: Any) -> Any:
+        data = copy.deepcopy(data)
+        if isinstance(data, dict):
+            files = data.pop("files")
+            # promote files list to dict if necessary
+            if isinstance(files, list):
+                # If files is a list, convert it to a dict and let it pass through the rest of the promotion logic
+                tmp = [f.rsplit(":", maxsplit=1) for f in files]
+                files = {}
+                for fsplit in tmp:
+                    # Need a valid split into file name and object path
+                    if len(fsplit) > 1:
+                        # but ensure we don't catch 'root://' and split that
+                        if fsplit[1].startswith("//"):
+                            # no object path
+                            files[":".join(fsplit)] = None
+                        else:
+                            # file name and object path
+                            files[fsplit[0]] = fsplit[1]
+            data["files"] = files
+        elif isinstance(data, DatasetSpec):
+            data = data.model_dump()
+        elif not isinstance(data, DatasetSpec):
+            raise ValueError(
+                "DatasetSpec expects a dictionary with a 'files' key or a DatasetSpec instance"
+            )
+        return data
+
+    @model_validator(mode="after")
+    def check_form(self) -> Self:
+        """Check the form can be decompressed, validate the format if manually specified, and then ."""
+        # check_form
+        if self.form is not None:
+            # If there's a form, validate we can decompress it into an awkward form
+            try:
+                import awkward
+
+                from coffea.util import decompress_form
+
+                _ = awkward.forms.from_json(decompress_form(self.form))
+            except Exception as e:
+                raise ValueError(
+                    "form: was not able to decompress_form into an awkward form"
+                ) from e
+
+        if self.format is None:
+
+            # set the format if not already set
+            union = set()
+            formats_by_name = {k: v.format for k, v in self.files.items()}
+            union.update(formats_by_name.values())
+            if len(union) == 1:
+                self.format = union.pop()
+            else:
+                self.format = "|".join(union)
+        else:
+            # validate the format, if present
+            if not IOFactory.valid_format(self.format):
+                raise ValueError(f"format: format must be one of {IOFactory._formats}")
+
+        return self
+
+    # @computed_field
+    # @property
+    def joinable(self) -> bool:
```

Comment:
`joinable` and `check_form` share a lot of code - would try to condense to shared parts and reuse.